### ISSUE TITLE
你好，我阅读了论文和论文的结构图发现SRU代码可能不太正确

### ISSUE THREAD

**Issue by Bewear777** [cheng-haha/ScConv#13]
我认为这样更贴合原文描述，请指正。
 def forward(self, x):
        gn_x = self.gn(x)
        w_gamma = F.softmax(self.gn.gamma, dim=0)
        reweigts = self.sigomid(gn_x * w_gamma)
        # Gate
        # info_mask = reweigts > self.gate_treshold
        # noninfo_mask = reweigts <= self.gate_treshold
        # x_1 = info_mask * reweigts * x
        # x_2 = noninfo_mask * reweigts * x
        w1 = torch.where(reweigts> self.gate_treshold, torch.ones_like(reweigts), reweigts)  # 大于门限值的设为1，否则保留原值
        w2 = torch.where(reweigts > self.gate_treshold, torch.zeros_like(reweigts), reweigts)  # 大于门限值的设为0，否则保留原值
        x_1 = w1 * x
        x_2 = w2 * x
        x = self.reconstruct(x_1, x_2)
        return x

**Comment by Sheeran2000** [comment]
我也感觉SRU有点问题 请问你对比了这两个代码吗

**Comment by Bewear777** [comment]
> 我也感觉SRU有点问题 请问你对比了这两个代码吗
作者论文对于gate部分的描述：We set those weights above the threshold to 1 to obtain the informative weights W1 while setting them to 0 to gain the non-informative weights W2(the threshold is set to 0.5 in the experiments).我认为可能是reweigts中高于阈值的置为1，但低于阈值的不变，得到W1；reweigts中低于阈值的置为0，但高于阈值的不变，得到W2；然后：Finally, we multiply input features X by W1 and W2respectively, yielding two weighted features: the informative ones Xw1 and less informative ones Xw2。就是将输入特征 X 分别乘以 W1 和 W2，得到两个加权特征。这里包括结构图应该是原X而不是处理后的GN(X)乘以权重。
这只是我个人的浅薄理解哈。

**Comment by Sheeran2000** [comment]
作者说不乘以GN(x), gn里的权重不变化  我也不懂为什么

**Comment by cheng-haha** [comment]
https://github.com/cheng-haha/ScConv/issues/8

**Comment by cheng-haha** [comment]
请你按照这里的issue：https://github.com/cheng-haha/ScConv/issues/8  测试一番，有什么其它可能的想法或者解决方案可以告知我，谢谢！

**Comment by cheng-haha** [comment]
已修正，
